src/ImageSharp.ruleset

```diff
@@ -1,4 +1,7 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <RuleSet Name="ImageSharp" ToolsVersion="17.0">
   <Include Path="..\shared-infrastructure\sixlabors.ruleset" Action="Default" />
+  <Rules AnalyzerId="Microsoft.CodeAnalysis.CSharp.NetAnalyzers" RuleNamespace="Microsoft.CodeAnalysis.CSharp.NetAnalyzers">
+    <Rule Id="CA2022" Action="Info" />
```

Comment:
For my learning, why was this analyzer downgraded from the default `Warning` to `Info`? The issues that this analyzer flags are typically bugs when reading from Streams.

Reply:
The warnings were in areas of our code where it wasn’t really applicable. Our decoders are designed to handle degenerate images and in each warning instance we had pre-allocated an empty buffer to read to. If we don’t get a full read then those buffers remain partially unfillled and the decoding process will return when we look for the next chunk.

Reply:
Hmm, I'm wondering if we have a missed use case / bug in our analyzer then. Do you happen to have a link to example code that it flagged? Maybe it just couldn't understand the pattern.

Reply:
I don't think the analyzer will ever be possible to follow the code well enough here.

Take this for example. In `ReadFrame` in the GIF decoder we attempt to read a local color palette from a stream. The analyzer will warn us that we might not get the full read though.

https://github.com/SixLabors/ImageSharp/blob/2d7cf48561765b2a4fc8db6c38c043d676ce5760/src/ImageSharp/Formats/Gif/GifDecoderCore.cs#L432

In this instance we simply do not care as the following method `ReadFrameColors` reads and checks the next byte. If it's invalid then we do nothing.

https://github.com/SixLabors/ImageSharp/blob/2d7cf48561765b2a4fc8db6c38c043d676ce5760/src/ImageSharp/Formats/Gif/GifDecoderCore.cs#L533-L534 

We also do a sense check and break when performing the next iteration to find additional frames so the decoder will quit after decoding as much information as possible.

https://github.com/SixLabors/ImageSharp/blob/2d7cf48561765b2a4fc8db6c38c043d676ce5760/src/ImageSharp/Formats/Gif/GifDecoderCore.cs#L155-L156

So, our code is safe, we could perhaps break earlier but we would end up complicating the code in order to do so. `Info` will work in our scenarios as it will prompt us to check the code but not inhibit our ability to work.

Reply:
Taking a quick look, I believe the reason why the code is safe is because `BufferedReadStream` guarantees it will fill out the entire passed in buffer on each call to `.Read`.

https://github.com/SixLabors/ImageSharp/blob/2d7cf48561765b2a4fc8db6c38c043d676ce5760/src/ImageSharp/IO/BufferedReadStream.cs#L176-L195

`ReadToBufferDirectSlow` and `ReadToBufferViaCopySlow` read from the underlying stream in a loop until it fills the buffer or hits the end of the stream.

https://github.com/SixLabors/ImageSharp/blob/2d7cf48561765b2a4fc8db6c38c043d676ce5760/src/ImageSharp/IO/BufferedReadStream.cs#L347-L357

https://github.com/SixLabors/ImageSharp/blob/2d7cf48561765b2a4fc8db6c38c043d676ce5760/src/ImageSharp/IO/BufferedReadStream.cs#L320-L326

https://github.com/SixLabors/ImageSharp/blob/2d7cf48561765b2a4fc8db6c38c043d676ce5760/src/ImageSharp/IO/BufferedReadStream.cs#L270-L288